File: app/infrastructure/encryption/rotation_journal.py
"""Crash-safe key-rotation journal.

Rotation re-encrypts every object under a new layer key, then rewraps the
header. If the process dies in the middle, some objects are under the new key
and the header still unwraps the old one. The journal stores the new key
wrapped with the old key (AEAD, same primitive as objects) plus the object ids
already rewritten, so the next unlock or rotate can finish the job.

The journal is ciphertext. It is useless without the old layer key, which still
comes from the password. It is deleted after the new header is committed.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

from app.infrastructure.encryption.primitives import decrypt, encrypt
from app.infrastructure.storage.paths import replace_atomic

JOURNAL_NAME = "rotation.journal"
AAD = b"strata-rotation-journal-v1"


@dataclass(frozen=True)
class RotationJournalState:
    layer_id: str
    manifest_object_id: str
    padding_enabled: bool
    done_object_ids: tuple[str, ...]
    new_key: bytes


class RotationJournal:
    def __init__(self, root: Path) -> None:
        self.path = root / JOURNAL_NAME

    def exists(self) -> bool:
        return self.path.is_file()

    def save(
        self,
        *,
        layer_id: str,
        manifest_object_id: str,
        padding_enabled: bool,
        done_object_ids: list[str],
        old_key: bytes,
        new_key: bytes,
    ) -> None:
        nonce, wrapped = encrypt(old_key, new_key, AAD)
        payload = {
            "v": 1,
            "layer_id": layer_id,
            "manifest_object_id": manifest_object_id,
            "padding_enabled": padding_enabled,
            "done_object_ids": done_object_ids,
            "nonce": nonce.hex(),
            "wrapped_new_key": wrapped.hex(),
        }
        temporary = self.path.with_name(self.path.name + ".tmp")
        try:
            temporary.write_text(json.dumps(payload, indent=2), encoding="utf-8")
            replace_atomic(temporary, self.path)
        except BaseException:
            temporary.unlink(missing_ok=True)
            raise

    def load(self, old_key: bytes) -> RotationJournalState:
        raw = json.loads(self.path.read_text(encoding="utf-8"))
        new_key = decrypt(
            old_key,
            bytes.fromhex(str(raw["nonce"])),
            bytes.fromhex(str(raw["wrapped_new_key"])),
            AAD,
        )
        done = raw.get("done_object_ids", [])
        if not isinstance(done, list):
            done = []
        return RotationJournalState(
            layer_id=str(raw["layer_id"]),
            manifest_object_id=str(raw["manifest_object_id"]),
            padding_enabled=bool(raw.get("padding_enabled", True)),
            done_object_ids=tuple(str(item) for item in done),
            new_key=new_key,
        )

    def clear(self) -> None:
        self.path.unlink(missing_ok=True)
        self.path.with_name(self.path.name + ".tmp").unlink(missing_ok=True)

File: dotnet/src/Strata.Infrastructure/Encryption/RotationJournal.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Strata.Core.Errors;
using Strata.Infrastructure.Storage;

namespace Strata.Infrastructure.Encryption;

/// <summary>The state a half-finished rotation left behind.</summary>
public sealed record RotationJournalState(
    string LayerId,
    string ManifestObjectId,
    bool PaddingEnabled,
    IReadOnlyList<string> DoneObjectIds,
    byte[] NewKey);

/// <summary>
/// Crash-safe key-rotation journal — port of
/// <c>app/infrastructure/encryption/rotation_journal.py</c>.
/// </summary>
/// <remarks>
/// Rotation re-encrypts every object under a new layer key, then rewraps the
/// header. If the process dies in the middle, some objects are under the new key
/// and the header still unwraps the old one. The journal stores the new key wrapped
/// with the old key (AEAD, same primitive as objects) plus the object ids already
/// rewritten, so the next unlock or rotate can finish the job.
/// <para>
/// The journal is ciphertext. It is useless without the old layer key, which still
/// comes from the password. It is deleted after the new header is committed.
/// </para>
/// </remarks>
public sealed class RotationJournal
{
    public const string JournalName = "rotation.journal";

    public static ReadOnlySpan<byte> Aad => "strata-rotation-journal-v1"u8;

    public RotationJournal(string root)
    {
        Path = System.IO.Path.Combine(root, JournalName);
    }

    public string Path { get; }

    private string TemporaryPath => Path + ".tmp";

    public bool Exists() => File.Exists(Path);

    public void Save(
        string layerId,
        string manifestObjectId,
        bool paddingEnabled,
        IReadOnlyList<string> doneObjectIds,
        ReadOnlySpan<byte> oldKey,
        ReadOnlySpan<byte> newKey)
    {
        var (nonce, wrapped) = AeadPrimitives.Encrypt(oldKey, newKey, Aad);
        var done = new JsonArray();
        foreach (var id in doneObjectIds)
        {
            done.Add(id);
        }

        var payload = new JsonObject
        {
            ["v"] = 1,
            ["layer_id"] = layerId,
            ["manifest_object_id"] = manifestObjectId,
            ["padding_enabled"] = paddingEnabled,
            ["done_object_ids"] = done,
            ["nonce"] = Convert.ToHexString(nonce).ToLowerInvariant(),
            ["wrapped_new_key"] = Convert.ToHexString(wrapped).ToLowerInvariant(),
        };

        AtomicFile.WriteText(Path, payload.ToJsonString(new JsonSerializerOptions { WriteIndented = true }));
    }

    public RotationJournalState Load(ReadOnlySpan<byte> oldKey)
    {
        JsonNode? node;
        try
        {
            node = JsonNode.Parse(File.ReadAllText(Path));
        }
        catch (Exception exc) when (exc is JsonException or IOException or UnauthorizedAccessException)
        {
            throw new DecryptionException("The rotation journal could not be read.", exc);
        }

        if (node is not JsonObject raw)
        {
            throw new DecryptionException("The rotation journal is not valid.");
        }

        byte[] newKey;
        try
        {
            newKey = AeadPrimitives.Decrypt(
                oldKey,
                Convert.FromHexString(raw["nonce"]!.GetValue<string>()),
                Convert.FromHexString(raw["wrapped_new_key"]!.GetValue<string>()),
                Aad);
        }
        catch (Exception exc) when (exc is NullReferenceException or InvalidOperationException
            or FormatException or ArgumentException)
        {
            throw new DecryptionException("The rotation journal is not valid.", exc);
        }

        var done = raw["done_object_ids"] is JsonArray array
            ? array.Select(item => item?.GetValue<string>() ?? string.Empty).ToList()
            : [];

        return new RotationJournalState(
            LayerId: raw["layer_id"]?.GetValue<string>() ?? string.Empty,
            ManifestObjectId: raw["manifest_object_id"]?.GetValue<string>() ?? string.Empty,
            PaddingEnabled: raw["padding_enabled"]?.GetValue<bool>() ?? true,
            DoneObjectIds: done,
            NewKey: newKey);
    }

    public void Clear()
    {
        AtomicFile.TryDelete(Path);
        AtomicFile.TryDelete(TemporaryPath);
    }
}
